examples/policies/multi-asset-policy/multi-asset-policy.pact

```diff
@@ -0,0 +1,359 @@
+    (enforce-guard ADMIN-KS))
+
+  (defconst OPERATOR-GUARD-MSG-KEY:string "operator_guard")
+  (defconst ASSETS-MSG-KEY:string "assets")
+
+  ; PROTOCOL CAPABILITIES
+
+  (defcap PROPOSE_ASSET (token-id:string uri:string)
+    (enforce-guard (at 'guard (read token-operators token-id)))
+  )
+  
+  (defcap REJECT_ASSET (token-id:string asset-id:integer owner:string)
+    (enforce-guard (marmalade-v2.ledger.account-guard token-id owner))
+  )  
+
+  (defcap REJECT_ALL_ASSETS (token-id:string owner:string)
+    (enforce-guard (marmalade-v2.ledger.account-guard token-id owner))
+  )
+  
+  (defcap ACCEPT_ASSET (token-id:string asset-id:integer owner:string)
+    (enforce-guard (marmalade-v2.ledger.account-guard token-id owner))
+  )
+
+  (defcap SET_ASSET_PRIORITY (token-id:string asset-id:integer priority:integer owner:string)
+    (enforce-guard (marmalade-v2.ledger.account-guard token-id owner))
+  )
+
+  ; PROTOCOL EVENTS
+
+  (defcap ASSET_PROPOSED (token-id:string asset-id:integer uri:string)
+    @doc "Emitted when new asset is proposed"
+    @event
+    true
+  )
+
+  (defcap ASSET_REJECTED (token-id:string asset-id:integer uri:string owner:string)
+    @doc "Emitted when the asset has been rejected"
+    @event
+    true
+  )
+    
+  (defcap ASSET_ACCEPTED (token-id:string asset-id:integer uri:string owner:string)
+    @doc "Emitted when the asset has been accepted"
+    @event
+    true
+  )
+
+  (defcap ASSET_SET (token-id:string asset-id:integer uri:string)
+    @doc "Emitted at init for fungible tokens"
+    @event
+
+    true
+  )
+
+  (defcap ASSET_PRIORITY_SET (token-id:string asset-id:integer priority:integer owner:string)
+    @doc "Emitted when new asset priority is set"
+    @event
+    true
+  )
+
+  (defschema token-asset-schema 
+    assets:[string]
+  )
+
+  (defschema token-operators-schema
+    guard:guard
+  )
+
+  (deftable token-assets:{token-asset-schema})
+  (deftable proposed-assets:{token-asset-schema})
+  (deftable token-operators:{token-operators-schema})
+
+  ; PROTOCOL FUNCTIONS
+
+  (defun get-asset:string (token-id:string asset-id:integer)
+    (at asset-id (at 'assets (read token-assets token-id)))
+  )  
+  
+  (defun get-assets:[string] (token-id:string)
+    (with-default-read token-assets token-id 
+      { 'assets: [] } 
+      { 'assets := assets } 
+      assets
+    )
+  )
+
+  (defun propose-asset:bool (token-id:string uri:string)
+    (enforce-non-fungible-policy token-id)
+    (enforce (not (= uri "")) "URI cannot be empty")
+    (let* (
+      (assets:[string] (+ (get-proposed-assets token-id) [uri]))
+      (asset-id:integer (- (length assets) 1))
+    )
+      (with-capability (PROPOSE_ASSET token-id uri)
+        (write proposed-assets token-id { 'assets: assets })
+        (emit-event (ASSET_PROPOSED token-id asset-id uri))
+        true
+      )
+    )
+  )
+
+  (defun replace-proposed-asset:bool (token-id:string asset-id:integer uri:string)
+    (enforce-non-fungible-policy token-id)
```

Comment:
`enforce-non-fungible-policy` is being used a lot of times but to improve gas usage for the asset proposal use case, the only place where this needs to be validated is at `propose-asset`. 
All the other functions rely on an asset being there.
Only for `replace-proposed-asset` a check might need to be added that ensures there is an asset to begin with, something like:

`(enforce (> (length assets) 0) "No assets found")`